File: packages/sdk/ruby/packages/admin/lib/edgebase_admin/push_client.rb
# frozen_string_literal: true

require "edgebase_core"

module EdgebaseAdmin
  # Client for push notification operations.
  #
  #   result = admin.push.send("userId", { "title" => "Hello", "body" => "World" })
  #   result = admin.push.send_many(["u1", "u2"], { "title" => "News" })
  #   logs = admin.push.get_logs("userId")
  class PushClient
    def initialize(http_client)
      @http = http_client
      @admin_core = EdgebaseAdmin::GeneratedAdminApi.new(http_client)
    end

    # Send a push notification to a single user's devices.
    def send(user_id, payload)
      @admin_core.push_send({ "userId" => user_id, "payload" => payload })
    end

    # Send a push notification to multiple users.
    def send_many(user_ids, payload)
      @admin_core.push_send_many({ "userIds" => user_ids, "payload" => payload })
    end

    # Send a push notification directly to a specific FCM token.
    def send_to_token(token, payload, platform: nil)
      body = { "token" => token, "payload" => payload }
      body["platform"] = platform if platform
      @admin_core.push_send_to_token(body)
    end

    # Get registered device tokens for a user.
    def get_tokens(user_id)
      res = @admin_core.get_push_tokens(query: { "userId" => user_id })
      res.is_a?(Hash) ? (res["items"] || []) : []
    end

    # Get push send logs for a user (last 24 hours).
    def get_logs(user_id, limit: nil)
      params = { "userId" => user_id }
      params["limit"] = limit.to_s if limit
      res = @admin_core.get_push_logs(query: params)
      res.is_a?(Hash) ? (res["items"] || []) : []
    end

    # Send a push notification to an FCM topic.
    def send_to_topic(topic, payload)
      @admin_core.push_send_to_topic({ "topic" => topic, "payload" => payload })
    end

    # Broadcast a push notification to all devices via /topics/all.
    def broadcast(payload)
      @admin_core.push_broadcast({ "payload" => payload })
    end
  end
end

File: packages/sdk/dart/packages/admin/lib/src/push_client.dart
/// PushClient — Push notification management for Admin SDK.
///
/// ```dart
/// final result = await client.push.send('userId', {'title': 'Hello', 'body': 'World'});
/// final result = await client.push.sendMany(['u1', 'u2'], {'title': 'News'});
/// final logs = await client.push.getLogs('userId');
/// ```
import 'package:edgebase_core/src/http_client.dart';
import 'generated/admin_api_core.dart';

/// Client for push notification operations.
class PushClient {
  late final GeneratedAdminApi _core;

  PushClient(HttpClient http) : _core = GeneratedAdminApi(http);

  /// Send a push notification to a single user's devices.
  Future<Map<String, dynamic>> send(
    String userId,
    Map<String, dynamic> payload,
  ) async {
    final res = await _core.pushSend({
      'userId': userId,
      'payload': payload,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Send a push notification to multiple users (no limit — server chunks internally).
  Future<Map<String, dynamic>> sendMany(
    List<String> userIds,
    Map<String, dynamic> payload,
  ) async {
    final res = await _core.pushSendMany({
      'userIds': userIds,
      'payload': payload,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Send a push notification directly to a specific FCM token.
  Future<Map<String, dynamic>> sendToToken(
    String token,
    Map<String, dynamic> payload, {
    String? platform,
  }) async {
    final res = await _core.pushSendToToken({
      'token': token,
      'payload': payload,
      if (platform != null) 'platform': platform,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Get registered device tokens for a user — token values NOT exposed.
  Future<List<Map<String, dynamic>>> getTokens(String userId) async {
    final res = await _core.getPushTokens({'userId': userId});
    if (res is Map<String, dynamic> && res['items'] is List) {
      return (res['items'] as List)
          .whereType<Map<String, dynamic>>()
          .toList();
    }
    return [];
  }

  /// Get push send logs for a user (last 24 hours).
  Future<List<Map<String, dynamic>>> getLogs(
    String userId, {
    int? limit,
  }) async {
    final query = <String, String>{'userId': userId};
    if (limit != null) query['limit'] = limit.toString();
    final res = await _core.getPushLogs(query);
    if (res is Map<String, dynamic> && res['items'] is List) {
      return (res['items'] as List)
          .whereType<Map<String, dynamic>>()
          .toList();
    }
    return [];
  }

  /// Send a push notification to an FCM topic.
  Future<Map<String, dynamic>> sendToTopic(
    String topic,
    Map<String, dynamic> payload,
  ) async {
    final res = await _core.pushSendToTopic({
      'topic': topic,
      'payload': payload,
    });
    return res is Map<String, dynamic> ? res : {};
  }

  /// Broadcast a push notification to all devices via /topics/all.
  Future<Map<String, dynamic>> broadcast(
    Map<String, dynamic> payload,
  ) async {
    final res = await _core.pushBroadcast({
      'payload': payload,
    });
    return res is Map<String, dynamic> ? res : {};
  }
}
